cli/azd/pkg/project/framework_service_docker.go

```diff
@@ -173,39 +178,32 @@ func (p *dockerProject) Package(
 	ctx context.Context,
 	serviceConfig *ServiceConfig,
 	buildOutput *ServiceBuildResult,
-) *async.TaskWithProgress[*ServicePackageResult, ServiceProgress] {
-	return async.RunTaskWithProgress(
-		func(task *async.TaskContextWithProgress[*ServicePackageResult, ServiceProgress]) {
-			imageId := buildOutput.BuildOutputPath
-			if imageId == "" {
-				task.SetError(errors.New("missing container image id from build output"))
-				return
-			}
+) (*ServicePackageResult, error) {
+	imageId := buildOutput.BuildOutputPath
+	if imageId == "" {
+		return nil, errors.New("missing container image id from build output")
+	}
 
-			localTag, err := p.containerHelper.LocalImageTag(ctx, serviceConfig)
-			if err != nil {
-				task.SetError(fmt.Errorf("generating local image tag: %w", err))
-				return
-			}
+	localTag, err := p.containerHelper.LocalImageTag(ctx, serviceConfig)
+	if err != nil {
+		return nil, fmt.Errorf("generating local image tag: %w", err)
+	}
 
-			// Tag image.
-			log.Printf("tagging image %s as %s", imageId, localTag)
-			task.SetProgress(NewServiceProgress("Tagging docker image"))
-			if err := p.docker.Tag(ctx, serviceConfig.Path(), imageId, localTag); err != nil {
-				task.SetError(fmt.Errorf("tagging image: %w", err))
-				return
-			}
+	// Tag image.
+	log.Printf("tagging image %s as %s", imageId, localTag)
+	p.publisher.Send(ctx, messaging.NewMessage(ProgressMessage, "Tagging docker image"))
```

Comment:
Service component can publish any type of message and consumers can subscribe to the types of messages they care about.